### ISSUE TITLE
[no-unused-vars] false-positive for class expressions

### ISSUE THREAD

**Issue by ocavue** [typescript-eslint/typescript-eslint#2831]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

 

```
$ git clone https://github.com/ocavue/typescript-eslint-issue-2831 /tmp/eslint-issue
$ cd /tmp/eslint-issue
$ yarn install 
$ yarn run lint 
```

**Expected Result**

No warning.

**Actual Result**

```
  2:11  warning  'C1' is defined but never used  @typescript-eslint/no-unused-vars
  4:11  warning  'C2' is defined but never used  @typescript-eslint/no-unused-vars
  6:11  warning  'C3' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

If I downgrade `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to version `4.8.2`, no warning occurs.
**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `15.2.1` |

**Comment by bradzacher** [comment]
that's an interesting pattern I've never seen before.

**Comment by scott-lc** [comment]
Expanding on this, 4.9 introduced a regression in this scenario:

```
// Inside of a namespace, `no-unused-vars` is incorrectly flagged.
export namespace Alpha {
   export interface Beta {
     doSomething: (value: unknown) => void;   // <--- Incorrect v4.9 `no-unused-vars` warning
  }
}

// The exact same thing outside of a namespace is fine.
export interface Beta {
   doSomething: (value: unknown) => void;   // All good
}
```

**Comment by bradzacher** [comment]
@scott-lc - unrelated. please file a new issue.

**Comment by derekparsons718** [comment]
I have a very similar case. Only difference is that the class is declared as an object property instead of an array element, and the object is exported for use in other files.

```typescript
export const objectWithClasses = {
    controller: class MyClass { //warning: 'MyClass' is defined but never used
        //...
    }
}
```

I assume this is the same problem, but I can open a new issue if desired.

**Comment by bradzacher** [comment]
@derekparsons718 - yup that's the same issue with class expressions.

**Comment by lukeapage** [comment]
Assume this is the same too:

```
const a = () => {
    return class A {};
};
```

**Comment by traverse** [comment]
> Assume this is the same too:
> 
> ```
> const a = () => {
>     return class A {};
> };
> ```

Yep that one as well, just ran into that exact issue.

**Comment by ocavue** [comment]
Just added the examples above into the origin issue description. In case someone else ran into the same cases.

**Comment by ocavue** [comment]
@bradzacher I can confirm that version 4.9.1 fixes this issue. Thanks for your help.

**Comment by lukeapage** [comment]
@bradzacher 
The example I posted is still broken in 4.9.1. New issue or will you re-open?
https://github.com/typescript-eslint/typescript-eslint/issues/2831#issuecomment-737891467

**Comment by ocavue** [comment]
> @bradzacher 
> 
> The example I posted is still broken in 4.9.1. New issue or will you re-open?
> 
> https://github.com/typescript-eslint/typescript-eslint/issues/2831#issuecomment-737891467

That is strange because I remember that I run through all the examples (including yours) and it seems to be okay. Can you check your lock files (e.g. yarn.lock) and see if there are two versions of this package.

**Comment by bradzacher** [comment]
Yup - working fine against master.
You likely have an old version installed.

**Comment by lukeapage** [comment]
Sorry! I made a mistake and had accidentally reverted the dependabot branch to 2.9.0
